uqcsbot/snailrace.py

```diff
@@ -0,0 +1,97 @@
+import discord, asyncio
+from discord import app_commands, ui
+
+from discord.ext import commands
+from uqcsbot.bot import UQCSBot
+
+import uqcsbot.utils.snailrace_utils as snail
+
+# Trying out Discord buttons for Snail Race Interactions
+class SnailRaceView(discord.ui.View):
+    def __init__(self, raceState: snail.SnailRaceState):
+        super().__init__(timeout=snail.SNAILRACE_OPEN_TIME)
+        self.raceState = raceState
+    
+    async def on_timeout(self):
+        """
+        Called when the view times out. This will deactivate the buttons and
+        begine the race.
+        """
+        for child in self.children:
+            child.disabled = True
+        await self.raceState.open_interaction.edit_original_response(content=snail.SNAILRACE_ENTRY_CLOSE, view=self)
+        await self.raceState.race_start()
+
+    @ui.button(label="Enter Race", style=discord.ButtonStyle.primary)
+    async def button_callback(self, interaction: discord.Interaction, button: discord.ui.Button):
+        if self.raceState.add_racer(interaction.user):
+            await interaction.response.send_message(snail.SNAILRACE_JOIN % interaction.user.mention)
+            return
+        
+        await interaction.response.send_message(snail.SNAILRACE_ALREADY_JOINED % interaction.user.mention)
+        
+
+class SnailRace(commands.Cog):
+    snailrace_group = app_commands.Group(name="snailrace", description="Snail Race Commands")
+
+    def __init__(self, bot: UQCSBot):
+        self.bot = bot
+        self.race = snail.SnailRaceState(self.start_racing)
+
+    @snailrace_group.command(name="open")
+    async def open_race(self, interaction: discord.Interaction):
+        """Open a new race for racers"""
+
+        # Check if there is a race on
+        if self.race.racing:
+            await interaction.response.send_message(snail.SNAILRACE_ENTRY_ERR)
+            return
+
+        # Open up a new race for racers
+        self.race.open_race(interaction)
+        await interaction.response.send_message(snail.SNAILRACE_ENTRY_MSG, view=SnailRaceView(self.race))
+
+    async def start_racing(self, interaction: discord.Interaction):
```

Comment:
After talking to @andrewj-brown, I think that it would be nice to remove some the state of the race within this function into `SnailRaceState`.

Reply:
I dont know what you mean by this, so I just moved the whole `start_racing` function to the `SnailRaceState` class. Please let me know what else you mean.

Reply:
To elaborate, I like the splitup of
- `SnailRaceState` handles game logic
- `SnailRace` handles cog I/O
- plus the other classes, and in different files

I'm just not a fan of passing start_racing as an argument to `SnailRaceState` being the way to handle that inter-class communication. I can't really think of better options though, the best paradigm I've got here is MVC from CSSE1001.